pim6d: Sometimes doesn't send PIM Register towards RP

### Description

A PIM IPv6 router which is not the RP sometimes doesn't send PIM Register packets towards the RP when a new MC traffic source becomes active. This seems to happen when the router receives an MLD join before the MC traffic starts. If the first packet of the actual MC traffic is received before the MLD join, PIM Register is sent towards the RP as expected.

Observed on a VRF-enabled PIM router. Not checked yet whether this also happens without VRF. Only happens with IPv6, IPv4 in the same test case works fine.

### Version

```text
Git master c288e5fbaf73bafb9aa15ade13969fc321e321db
```

### How to reproduce

Use latest version of `pim_igmp_vrf` topotest from https://github.com/FRRouting/frr/pull/18329. Run topotest using `pytest pim_igmp_vrf`. In about 10% of the cases, this will fail with `FAILED pim_igmp_vrf/test_pim_vrf.py::test_mcast_vrf_blue - AssertionError: PIM IPv4 R1 VRF blue: No PIM Register sent towards RP` (could also be VRF red).

To make it fail reliably, change topotest as follows:
```diff
diff --git a/tests/topotests/pim_igmp_vrf/test_pim_vrf.py b/tests/topotests/pim_igmp_vrf/test_pim_vrf.py
index 9615102fe..0327a54d0 100755
--- a/tests/topotests/pim_igmp_vrf/test_pim_vrf.py
+++ b/tests/topotests/pim_igmp_vrf/test_pim_vrf.py
@@ -108,6 +108,7 @@ import re
 import sys
 import pytest
 import logging
+import time
 
 # Save the Current Working Directory to find configuration files.
 CWD = os.path.dirname(os.path.realpath(__file__))
@@ -451,8 +452,9 @@ def check_mcast46_entry(
     )
 
     with McastTesterHelper(tgen) as helper:
-        helper.run(sender, ["--send=0.7", mcastaddr, str(sender) + "-eth0"])
         helper.run(receiver, [mcastaddr, str(receiver) + "-eth0"])
+        time.sleep(1)
+        helper.run(sender, ["--send=0.7", mcastaddr, str(sender) + "-eth0"])
 
         logger.info("mcast join and source for {} started".format(mcastaddr))
 
```

### Expected behavior

PIM Register always sent towards RP when MC traffic starts (only one, since the RP should with PIM Register-Stop), test case passes

### Actual behavior

Depending on 1st packet received at router (MLD join or actual MC traffic), no PIM Register is sent

### Additional context

_No response_

### Checklist

- [x] I have searched the open issues for this bug.
- [x] I have not included sensitive information in this report.

--- a/lib/if.c
+++ b/lib/if.c
@@ -604,9 +604,9 @@ struct connected *if_lookup_address(const void *matchaddr, int family,
 
 	FOR_ALL_INTERFACES (vrf, ifp) {
 		frr_each (if_connected, ifp->connected, c) {
-			if (c->address && (c->address->family == AF_INET)
-			    && prefix_match(CONNECTED_PREFIX(c), &addr)
-			    && (c->address->prefixlen > bestlen)) {
+			if (c->address && (c->address->family == family) &&
+			    prefix_match(CONNECTED_PREFIX(c), &addr) &&
+			    (c->address->prefixlen > bestlen)) {
 				bestlen = c->address->prefixlen;
 				match = c;
 			}

--- a/pimd/pim_mroute.c
+++ b/pimd/pim_mroute.c
@@ -436,30 +436,53 @@ int pim_mroute_msg_wholepkt(int fd, struct interface *ifp, const char *buf,
 	up = pim_upstream_find(pim_ifp->pim, &sg);
 	if (!up) {
 		pim_sgaddr star = sg;
+
 		star.src = PIMADDR_ANY;
 
 		up = pim_upstream_find(pim_ifp->pim, &star);
 
 		if (up && PIM_UPSTREAM_FLAG_TEST_CAN_BE_LHR(up->flags)) {
-			up = pim_upstream_add(pim_ifp->pim, &sg, ifp,
-					      PIM_UPSTREAM_FLAG_MASK_SRC_LHR,
-					      __func__, NULL);
-			if (!up) {
+			struct connected *src_conn = NULL;
+			struct pim_interface *src_pim_ifp = NULL;
+
+			/*
+			 * Now let's test to see if this also can be
+			 * a FHR as well, As that this is a possibility
+			 * and if it is we must treat it as such.
+			 */
+			src_conn = if_lookup_address(&sg.src, PIM_AF, pim_ifp->pim->vrf->vrf_id);
+
+			if (src_conn && src_conn->ifp->info)
+				src_pim_ifp = src_conn->ifp->info;
+
+			if (src_conn != NULL && src_pim_ifp && src_pim_ifp->pim_enable) {
+				up = pim_upstream_add(pim_ifp->pim, &sg, src_conn->ifp,
+						      PIM_UPSTREAM_FLAG_MASK_FHR, __func__, NULL);
+				PIM_UPSTREAM_FLAG_SET_SRC_STREAM(up->flags);
+				pim_upstream_keep_alive_timer_start(up,
+								    pim_ifp->pim->keep_alive_time);
+				pim_upstream_inherited_olist(pim_ifp->pim, up);
+				pim_upstream_update_join_desired(pim_ifp->pim, up);
+				if (!pim_is_group_filtered(pim_ifp, &sg.grp, &sg.src))
+					pim_register_join(up);
 				if (PIM_DEBUG_MROUTE)
-					zlog_debug(
-						"%s: Unable to create upstream information for %pSG",
-						__func__, &sg);
+					zlog_debug("%s: Treat WHOLEPKT on pimreg as FHR for %pSG",
+						   __func__, &sg);
+				goto have_up;
+			} else {
+				up = pim_upstream_add(pim_ifp->pim, &sg, ifp,
+						      PIM_UPSTREAM_FLAG_MASK_SRC_LHR, __func__,
+						      NULL);
+				pim_upstream_keep_alive_timer_start(up,
+								    pim_ifp->pim->keep_alive_time);
+				pim_upstream_inherited_olist(pim_ifp->pim, up);
+				pim_upstream_update_join_desired(pim_ifp->pim, up);
+
+				if (PIM_DEBUG_MROUTE)
+					zlog_debug("%s: Creating %s upstream on LHR", __func__,
+						   up->sg_str);
 				return 0;
 			}
-			pim_upstream_keep_alive_timer_start(
-				up, pim_ifp->pim->keep_alive_time);
-			pim_upstream_inherited_olist(pim_ifp->pim, up);
-			pim_upstream_update_join_desired(pim_ifp->pim, up);
-
-			if (PIM_DEBUG_MROUTE)
-				zlog_debug("%s: Creating %s upstream on LHR",
-					   __func__, up->sg_str);
-			return 0;
 		}
 		if (PIM_DEBUG_MROUTE_DETAIL) {
 			zlog_debug(
@@ -468,6 +491,8 @@ int pim_mroute_msg_wholepkt(int fd, struct interface *ifp, const char *buf,
 		}
 		return 0;
 	}
+
+have_up:
 
 	if (!up->rpf.source_nexthop.interface) {
 		if (PIM_DEBUG_PIM_TRACE)

--- a/pimd/pim_upstream.c
+++ b/pimd/pim_upstream.c
@@ -2340,6 +2340,23 @@ static bool pim_upstream_sg_running_proc(struct pim_upstream *up)
 			PIM_UPSTREAM_FLAG_SET_SRC_STREAM(up->flags);
 			pim_upstream_fhr_kat_start(up);
 		}
+
+		/*
+		 * Let's ensure that when we have an active source that we do not have any
+		 * register state for that is a FHR, that we allow the registration to
+		 * happen if it should be
+		 */
+		if (pim_upstream_could_register(up) && !pim_is_grp_ssm(pim, up->sg.grp) &&
+		    up->reg_state == PIM_REG_NOINFO && !event_is_scheduled(up->t_rs_timer) &&
+		    !PIM_UPSTREAM_DM_TEST_INTERFACE(up->flags) && pim->regiface &&
+		    pim->regiface->configured) {
+			if (PIM_DEBUG_PIM_TRACE)
+				zlog_debug("%s: add pimreg to %s[%s]", __func__, up->sg_str,
+					   pim->vrf->name);
+			PIM_UPSTREAM_FLAG_SET_FHR(up->flags);
+			pim_register_join(up);
+			pim_upstream_update_use_rpt(up, true /*update_mroute*/);
+		}
 		pim_upstream_keep_alive_timer_start(up, pim->keep_alive_time);
 		rv = true;
 	} else if (PIM_UPSTREAM_FLAG_TEST_SRC_LHR(up->flags)) {

